.github/workflows/CICD.yml

```diff
@@ -1,18 +1,148 @@
 name: CICD
+on: [pull_request, push]
+jobs:
+  test-lib:
+    name: Test library
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [macos-latest, ubuntu-latest, windows-latest]
+        toolchain: [stable, nightly]
+    continue-on-error: true
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v2
+      - name: Setup Rust
+        run: |
+          rustup toolchain add ${{ matrix.toolchain }} --component llvm-tools-preview
+          rustup override set ${{ matrix.toolchain }}
+      - name: Configure cache
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+            target/
+          key: test-${{ runner.os }}-cargo-${{ matrix.toolchain }}-${{ hashFiles('**/Cargo.lock') }}
+      - name: Test
+        run: cargo test --lib
+  test-coverage:
+    name: Test coverage (gcc-${{ matrix.gcc }}, clang-${{ matrix.clang }})
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        include:
+          - gcc: 5
+            clang: 7
+          - gcc: 6
+            clang: 7
+          - gcc: 7
+            clang: 7
+          - gcc: 7
+            clang: 8
+          - gcc: 7
+            clang: 9
+          - gcc: 7
+            clang: 11
+          - gcc: 10
+            clang: 11
+    env:
+      CXX: clang++-${{ matrix.clang }}
+      CLANG_CXX: clang++-${{ matrix.clang }}
+      GCC_CXX: g++-${{ matrix.gcc }}
+      GCOV: gcov-${{ matrix.gcc }}
+    continue-on-error: true
+    steps:
+      - name: Prepare C toolchains
+        run: |
+          # Add LLVM keys
+          curl -o- https://apt.llvm.org/llvm-snapshot.gpg.key | sudo apt-key add -
 
-on: [push, pull_request]
-
-env:
-  CARGO_TERM_COLOR: always
+          # Add extra APT repos
+          cat > /tmp/extra.list <<-EOF
+            deb http://archive.ubuntu.com/ubuntu/ bionic main restricted
+            deb http://archive.ubuntu.com/ubuntu/ bionic universe
 
-jobs:
-  build:
+            deb http://archive.ubuntu.com/ubuntu/ xenial main restricted
+            deb http://archive.ubuntu.com/ubuntu/ xenial universe
+            deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-7 main
+            deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-8 main
+            deb http://apt.llvm.org/focal/ llvm-toolchain-focal-9 main
+            deb http://apt.llvm.org/focal/ llvm-toolchain-focal-10 main
+            deb http://apt.llvm.org/focal/ llvm-toolchain-focal-11 main
+            deb http://apt.llvm.org/focal/ llvm-toolchain-focal-12 main
+            deb http://apt.llvm.org/focal/ llvm-toolchain-focal-13 main
+          EOF
+          sudo mv /tmp/extra.list /etc/apt/sources.list.d/
 
+          # Install toolchains
+          sudo apt-get update
+          sudo apt-get install -y clang-${{ matrix.clang }} g++-${{ matrix.gcc }}
```

Comment:
Disclaimer: I don't know anything about github actions, so in advance, sorry for the stupid questions.

Are those packages cached at some point ? If not, is it possible to cache them ?

Reply:
There is a caching action to do this, but we have to manually define the folders and files to cache.

Currently, it's already set up for Rust but not for APT. I found some community made actions that automate caching of APT, but according to @marco-c they're not permitted in Mozilla repos.

I'm not an expert on APT's structure, but a quick search yielded that the repo lists are cached at `/var/lib/apt/lists` and the DEBs are cached at `/var/cache/apt/archives`.

We might just give it a try and throw these in a GitHub Actions cache. Just not sure if that's enough, as `/var/cache/apt` contains some extra cache files, and I'm not sure about their relation to the cached DEBs.

Also, if we cache these files, how do we evict the cache? For Cargo, it's simple as we invalidate the cache whenever the `Cargo.lock` file changes, but for APT it seems more difficult as we don't have a lock file.

Reply:
Let's file this as a follow-up.